### PR TITLE
Adding the bulk status API call of the FHIR spec

### PR DESCRIPTION
**Pull request by adaykin**
Adding the bulk status API call of the FHIR spec

Adding the bulk status API call and returning appropriate status codes so that the client can have an understanding of what the status of the job is. Adding exception handling for different sorts of errors so that there will be less boilerplate code and that it's easier to map the types of exceptions to status codes. 

### Checklist

- [x] Tests and linting pass
- [x] Code checked for PHI/PII exposure

### Security
N/A

### CODE REVIEW

File: api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java [hunk 1]
```diff
@@ -109,11 +120,33 @@
...
+            throw new TooManyRequestsException("You are polling too frequently");
+        }
+
+        job.setLastPollTime(now);
+        jobService.updateJob(job);
+
         HttpHeaders responseHeaders = new HttpHeaders();
-        responseHeaders.add("X-Progress", "0%");
-        return new ResponseEntity<>(null, responseHeaders, HttpStatus.ACCEPTED);
+        if (job.getProgress().equals(100)) {
```
Comment (dkrylovsb) [inline api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java]: We should be checking the `status` field.

File: api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java [hunk 2]
```diff
@@ -109,11 +120,33 @@
...
+            responseHeaders.add("Expires", DateUtil.formatLocalDateTimeAsUTC(job.getExpires()));
+            JobCompletedResponse resp = new JobCompletedResponse();
+            resp.setTransactionTime(new DateTimeType(convertLocalDateTimeToDate(job.getCompletedAt())).toHumanDisplay());
+            resp.setRequest(job.getRequestURL());
+            resp.setOutput(job.getJobOutput().stream().filter(o -> !o.isError()).map(o -> new JobCompletedResponse.Output(o.getFhirResourceType(), o.getFilePath())).collect(Collectors.toList()));
+            resp.setError(job.getJobOutput().stream().filter(o -> o.isError()).map(o -> new JobCompletedResponse.Output(o.getFhirResourceType(), o.getFilePath())).collect(Collectors.toList()));
+            return new ResponseEntity<>(new ObjectMapper().valueToTree(resp), responseHeaders, HttpStatus.OK);
```
Comment (dkrylovsb) [inline api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java]: We will need to set `requiresAccessToken` to `true` as well.

File: api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java [hunk 3]
```diff
@@ -109,11 +120,33 @@
...
+        } else {
```
Comment (dkrylovsb) [inline api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java]: We should return this _only_ if the job is in `SUBMITTED` or `IN_PROGRESS` status

File: api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java [hunk 4]
```diff
@@ -109,11 +120,33 @@
...
+            responseHeaders.add("X-Progress", job.getProgress() + "% complete");
+            responseHeaders.add("Retry-After", Integer.toString(retryAfterDelay));
+            return new ResponseEntity<>(null, responseHeaders, HttpStatus.ACCEPTED);
+        }
```
Comment (dkrylovsb) [inline api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java]: Need to handle the `FAILED` job status and return 500 with an OperationOutcome as described in the spec, e.g.
```
Status: 500 Internal Server Error
Content-Type: application/json

{
 "resourceType": "OperationOutcome",
 "id": "1",
 "issue": [
  {
   "severity": "error",
   "code": "processing",
   "details": {
    "text": "An internal timeout has occurred"
   }
  }
 ]
}```

File: api/src/main/java/gov/cms/ab2d/api/controller/ErrorHandler.java [hunk 5]
```diff
@@ -19,12 +21,41 @@
...
+        return generateFHIRError(e, HttpStatus.INTERNAL_SERVER_ERROR);
+    }
+
+    // Most errors that are the fault of the client
+    @ExceptionHandler(IllegalArgumentException.class)
     public ResponseEntity<JsonNode> assertionException(final Exception e) throws IOException {
+        return generateFHIRError(e, HttpStatus.BAD_REQUEST);
+    }
+
+    @ExceptionHandler(EntityNotFoundException.class)
```
Comment (dkrylovsb) [inline api/src/main/java/gov/cms/ab2d/api/controller/ErrorHandler.java]: This is JPA-specific exception, which makes the service class a somewhat "leaky" abstraction, i.e. persistence-related stuff leaks to controllers. Let's just introduce our own exception class, just like you did with TooManyRequests: e.g. `ResourceNotFoundException`

File: api/src/main/java/gov/cms/ab2d/api/repository/JobRepository.java [hunk 6]
```diff
@@ -7,4 +7,5 @@
 @Repository
```
Comment (dkrylovsb) [inline api/src/main/java/gov/cms/ab2d/api/repository/JobRepository.java]: I just realized that this repository is in the `api` module. This needs to be moved to the `domain` module, since both the API and Worker will use this repository.

File: api/src/main/java/gov/cms/ab2d/api/service/JobService.java [hunk 7]
```diff
@@ -5,4 +5,8 @@
 public interface JobService {
```
Comment (dkrylovsb) [inline api/src/main/java/gov/cms/ab2d/api/service/JobService.java]: Same. Needs to go into `domain`

File: api/src/main/java/gov/cms/ab2d/api/service/JobServiceImpl.java [hunk 8]
```diff
@@ -36,4 +36,17 @@ public Job createJob(String resourceTypes, String url) {
```
Comment (dkrylovsb) [inline api/src/main/java/gov/cms/ab2d/api/service/JobServiceImpl.java]: Also into `domain`

Comment (dkrylovsb) [reply api/src/main/java/gov/cms/ab2d/api/service/JobServiceImpl.java]: In that case we would have a design problem. JobService won't need to access any API- or Worker-specific services.

File: api/src/main/java/gov/cms/ab2d/api/repository/JobRepository.java (hunk 6 above)
Comment (adaykin) [reply api/src/main/java/gov/cms/ab2d/api/repository/JobRepository.java]: Discussed over slack and this can be completed in a separate jira to refactor code to go into the shared module.

File: api/src/test/java/gov/cms/ab2d/api/controller/BulkDataAccessAPIIntegrationTests.java [hunk 9]
```diff
@@ -98,4 +106,85 @@ public void testPatientExportWithInvalidOutputFormat() throws Exception {
...
+
+        jobRepository.saveAndFlush(job);
+
+        this.mockMvc.perform(get(statusUrl).contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().is(500))
+                .andExpect(jsonPath("$.resourceType", Is.is("OperationOutcome")))
+                .andExpect(jsonPath("$.issue[0].severity", Is.is("error")))
+                .andExpect(jsonPath("$.issue[0].code", Is.is("invalid")))
+                .andExpect(jsonPath("$.issue[0].details.text", Is.is("JobProcessingException: Job failed while processing")));
+    }
```
Comment (dkrylovsb) [inline api/src/test/java/gov/cms/ab2d/api/controller/BulkDataAccessAPIIntegrationTests.java]: Need a test for 404 (job not found). Also need a test for an empty Job ID passed in the URl

Comment (adaykin) [reply api/src/test/java/gov/cms/ab2d/api/controller/BulkDataAccessAPIIntegrationTests.java]: Added in those tests
